### DISCUSSION TITLE
JOLI method error for operator

### DISCUSSION THREAD

**Discussion by tjk9501** [slimgroup/JUDI.jl#240] [Q&A] [answered]
┌ Info: JOLI method error for operator:
│   op_name = "judiJacobian{Float32, :born, judiDataSourceModeling{Float32, :forward}}"
└   op_type = judiJacobian{Float32, :born, judiDataSourceModeling{Float32, :forward}}
ERROR: LoadError: JOLI.joUtilsException("-(jo) not implemented")
Stacktrace:
 [1] jo_method_error(O::judiJacobian{Float32, :born, judiDataSourceModeling{Float32, :forward}}, s::String)
   @ JOLI ~/.julia/packages/JOLI/S3LJy/src/joUtils.jl:45
 [2] -(A::judiJacobian{Float32, :born, judiDataSourceModeling{Float32, :forward}})
   @ JOLI ~/.julia/packages/JOLI/S3LJy/src/joAbstractOperator/base_functions.jl:146
 [3] -(A::judiJacobian{Float32, :born, judiDataSourceModeling{Float32, :forward}}, B::judiJacobian{Float32, :born, judiDataSourceModeling{Float32, :forward}})
   @ JOLI ~/.julia/packages/JOLI/S3LJy/src/joAbstractLinearOperator/base_functions.jl:474
 [4] top-level scope
   @ /media/tju/0FD709D40FD709D4/jktong/Pipe/PipeFWIInversion/Pipe_2D_GN_Coher.jl:106

**Comment by mloubout** [comment]
Hi

Please provide some context and, if possible, a reproducer. This error message alone points to some script (/Pipe_2D_GN_Coher.jj) without any other information isn't enough to offer an answer.

**Reply by mloubout** [reply, marked as answer]
This might not be supported at the time. 

I'm also not sure this can actually work because you will need to compute `(J[isrc]+J[jsrc])'*(d_pred - d_obs[isrc])` which cannot work because the geometry of `d_pred - d_obs[isrc]` will not match the one of `J[isrc]+J[jsrc]`

**Comment by tjk9501** [comment]
I assume the length of isrc and jsrc are the same. If it is not supported, is it possible to add such operations in future  release?



---- Replied Message ----
| From | Mathias ***@***.***> |
| Date | 04/03/2024 11:51 |
| To | ***@***.***> |
| Cc | ***@***.***>***@***.***> |
| Subject | Re: [slimgroup/JUDI.jl] JOLI method error for operator (Discussion #240) |

This might not be supported at the time.

I'm also not sure this can actually work because you will need to compute (J[isrc]+J[jsrc])'*(d_pred - d_obs[isrc]) which cannot work because the geometry of d_pred - d_obs[isrc] will not match the one of J[isrc]+J[jsrc]

**Reply by mloubout** [reply]
Having the same length is not enough. When you are computing `J[i]'*d[i]` (or `J[i]*dm - d[i]`), and this is what you are doing, then the actual geometry need to match. You are solving a PDE for a given source and receiver position

Even before that `d_pred - d_obs[isrc]` is probably not correct because you are comparing two datasets that may not be located in the same area.

**Comment by tjk9501** [comment]
If J[isrc] is an actual matrix it will be much easier to do that. Anyway, please notify me if that is implemented in futute release.

Thanks!



---- Replied Message ----
| From | Mathias ***@***.***> |
| Date | 04/03/2024 12:00 |
| To | ***@***.***> |
| Cc | ***@***.***>***@***.***> |
| Subject | Re: [slimgroup/JUDI.jl] JOLI method error for operator (Discussion #240) |

Having the same length is not enough. When you are computing J[i]'*d[i] (or J[i]*dm - d[i]), and this is what you are doing, then the actual geometry need to match. You are solving a PDE for a given source and receiver position

Even before that d_pred - d_obs[isrc] is probably not correct because you are comparing two datasets that may not be located in the same area.

**Reply by mloubout** [reply]
This is not about being a matrix. `J[isrc]' *d[jsrc]` will not be supported as this is not a valid operation. `J[isrc]` is `J(Pr[isrc]*F[isrc]*Ps[isrc]'` while `d[jsrc]` is `Pr[jsrc]*F[jsrc]*Ps[jsrc]' and are not compatible. 

It is straightforward to (add) support valid operations such block diagonal operators i.e `[J[isrc] 0; 0 J[jsrc]]` which can be used to work on vectors like `[d[isrc]; d[jsrc]]`, or additions with valid geometries but `J[isrc]+J[jsrc]` cannot be supported.

**Comment by tjk9501** [comment]
Now I understand that you mean it is impossible to implement something like J[isrc]*d[jsrc]. Is it possible to return the values of J[isrc] and d[jsrc] as normal matrix array and vector Arrays? As both of them can be explicitly calculated.



---- Replied Message ----
| From | Mathias ***@***.***> |
| Date | 04/03/2024 12:09 |
| To | ***@***.***> |
| Cc | ***@***.***>***@***.***> |
| Subject | Re: [slimgroup/JUDI.jl] JOLI method error for operator (Discussion #240) |

This is not about being a matrix. J[isrc]' *d[jsrc] will not be supported as this is not a valid operation. J[isrc] is J(Pr[isrc]*F[isrc]*Ps[isrc]' while d[jsrc] is `Pr[jsrc]*F[jsrc]*Ps[jsrc]' and are not compatible.

It is straightforward to (add) support valid operations such block diagonal operators i.e [J[isrc] 0; 0 J[jsrc]] which can be used to work on vectors like [d[isrc]; d[jsrc]], or additions with valid geometries but J[isrc]+J[jsrc] cannot be supported.

**Reply by mloubout** [reply]
for `d[jsrc]` yes you can just do `vec` but for `J`, no it's not possible this would be too expensive and big to be formed as a dense matrix. For example a very tiny 2d problem (128x128 model, 10 sources, 121 receivers, 501 time samples) would already be 40Gb of memory so it's not realistic to build the matrix.

If you "really" want to build it then you would have to build it column by column by doing `J[;, i] = j*e_i` where `e_i` is the ith unit vector (all zero and 1 for row i). That would cost a lot t do as well

**Reply by tjk9501** [reply]
OK, I'll test it carefully.
